tests/support/client.rs

```diff
@@ -245,40 +271,135 @@ impl hyper::client::connect::Connect for Conn {
     type Error = ::std::io::Error;
     fn connect(&self, _: hyper::client::connect::Destination) -> Self::Future {
         let connected = hyper::client::connect::Connected::new().proxy(self.absolute_uris);
-        Box::new(self.connect_().map(|t| (t, connected)))
+        Box::new(self.initialize_connection().map(|t| (t, connected)))
     }
 }
 
-/// A wrapper around a TcpStream, allowing us to signal when the connection
-/// is dropped.
-struct RunningIo {
-    inner: TcpStream,
-    /// When this drops, the related Receiver is notified that the connection
-    /// is closed.
-    running: oneshot::Sender<()>,
+pub enum ConnectorFuture<F: Future> {
+    Init {
+        future: F,
+        tls: Option<(Arc<ClientConfig>, DNSName)>,
+        running: Arc<Mutex<Option<oneshot::Sender<()>>>>,
+    },
+    Handshake {
+        future: tokio_rustls::Connect<F::Item>,
+        running: Arc<Mutex<Option<oneshot::Sender<()>>>>,
+    },
+}
+
+impl<F> Future for ConnectorFuture<F>
+where
+    F: Future,
+    F::Item: AsyncRead + AsyncWrite + Send + 'static,
+    F::Error: From<io::Error>,
+{
+    type Item = RunningIo<F::Item>;
+    type Error = F::Error;
+
+    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+        fn take_running(r: &Mutex<Option<oneshot::Sender<()>>>) -> oneshot::Sender<()> {
```

Comment:
Take it or leave it: if this was a closure rather than a function, it could just capture `running`...